catalog/ui/catalog-ui-search/src/main/java/org/codice/ddf/catalog/ui/forms/builder/AttributeValueNormalizer.java

```diff
@@ -0,0 +1,109 @@
+import java.time.Instant;
+import java.time.format.DateTimeParseException;
+import java.util.Objects;
+import java.util.Optional;
+import javax.annotation.Nullable;
+import org.codice.ddf.catalog.ui.forms.filter.FilterProcessingException;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+class AttributeValueNormalizer {
+  private static final Logger LOGGER = LoggerFactory.getLogger(AttributeValueNormalizer.class);
+
+  private final AttributeRegistry registry;
+
+  AttributeValueNormalizer(AttributeRegistry registry) {
+    this.registry = registry;
+  }
+
+  String normalizeForJson(String property, String value) {
+    if (isNotNormalizableDateValue(property, value)) {
+      return value;
+    }
+    Instant instant = instantFromEpoch(value);
+    if (instant != null) {
+      return Objects.toString(instant);
+    }
+    return value;
+  }
+
+  String normalizeForXml(String property, String value) {
+    if (isNotNormalizableDateValue(property, value)) {
+      return value;
+    }
+    Instant epoch = instantFromEpoch(value);
+    if (epoch != null) {
+      return value;
+    }
+    Instant iso = instantFromIso(value);
+    if (iso != null) {
+      return Objects.toString(iso.toEpochMilli());
+    }
+    throw new FilterProcessingException("Unexpected date format on search form: " + value);
+  }
+
+  // Make this a 'private boolean isDateValue' and extract boolean functions for date checks as well
+  // Input (normalizeForXml) should allow exceptions to propagate
+  // Output (normalizeForJson) should be more forgiving so the feature is hardened against failure
+  private boolean isNotNormalizableDateValue(String property, String value) {
+    if (property == null || value == null) {
+      LOGGER.trace("Property [{}] or value [{}] was null, ignoring normalization", property, value);
+      return true;
+    }
+    Optional<AttributeDescriptor> optional = registry.lookup(property);
+    if (!optional.isPresent()) {
+      LOGGER.trace("No descriptor available for property [{}]", property);
+      return true;
+    }
+    AttributeDescriptor descriptor = optional.get();
+    if (!AttributeType.AttributeFormat.DATE.equals(descriptor.getType().getAttributeFormat())) {
+      LOGGER.trace(
+          "Descriptor for property [{}] and value [{}] did not have an AttributeFormat of DATE",
+          property,
+          value);
+      return true;
+    }
+    LOGGER.trace("Found date value [{}] for property [{}]", value, property);
+    return false;
+  }
+
+  @Nullable
+  private static Instant instantFromEpoch(String epochString) {
+    try {
+      long epoch = Long.parseLong(epochString);
+      return Instant.ofEpochMilli(epoch);
+    } catch (DateTimeException | NumberFormatException e) {
+      LOGGER.debug("Error parsing epoch string [{}]", epochString, e);
+      return null;
+    }
+  }
+
+  @Nullable
+  private static Instant instantFromIso(String isoString) {
+    try {
+      return Instant.parse(isoString);
```

Comment:
Just beware the caveats with Instant.parse https://docs.oracle.com/javase/8/docs/api/java/time/Instant.html#parse-java.lang.CharSequence-
and the associated https://docs.oracle.com/javase/8/docs/api/java/time/format/DateTimeFormatter.html#ISO_INSTANT
Its parsing a very specific iso8601 variant and does not support _all_ of them if that is your intent with a more defensive date parsing.

Reply:
Nice catch. I'll look into the facilities that endpoint util provides.

Reply:
So I checked out the parsing function in endpoint util and it turns out I need different semantics than what it offers. So I just added support for offsets to my code without adopting the util to get the functionality I needed.